### PR TITLE
Database: Perform one database action for ClusterTx methods.

### PR DESCRIPTION
**Pull request by masnax**
Finally getting this out there...

* Basic gist: Each method on `ClusterTx` will now (for the most part) only perform one database action. As a result, neither `UsedBy` fields nor URL generation are in the `db` package anymore. Instead, they are generated when handling API hits.

* Reference fields like `instance.Config`, `instance.Devices`, `instance.Profiles` have been removed from the database structs. These fields now have their own methods (i.e. `UpdateInstanceConfig, UpdateInstanceDevices`) and cannot be expected to be handled by `tx.{Get,Update,Create,Delete}Instance`. 
   * The generator now creates standalone methods for CRUDing reference fields.
   * These entities' `ToAPI` methods now have a `ClusterTx` argument, so that these fields will be generated for the API structs, which is when we will be needing them. 
   * In the case of `InstanceArgs` and `InstanceSnapshots`, `InstanceArgs` will hold all shared data between instances and snapshots, and will be passed around (Primarily in `InstanceList`) as a way to handle instances and snapshots uniformly.

* In several places, transactions are now held open longer or passed around as an argument so that we do not open new ones unnecessarily.

### CODE REVIEW

**Comment by stgraber** [comment]
Documentation preview available at: https://linuxcontainers.org/lxd/docs/pr.9775

**Comment by stgraber** [comment]
@masnax quick rebase please

**Comment by tomponline** [comment]
@stgraber are you OK if I wait until I get back to look at this one so I can give it sufficient attention? Thanks

**Comment by stgraber** [comment]
@tomponline if you'd like to review it, sure, otherwise I was going to bite the bullet and go through all those commits ;)

**Comment by tomponline** [comment]
> @tomponline if you'd like to review it, sure, otherwise I was going to bite the bullet and go through all those commits ;)

Absolutely happy for you to take a look too, in fact given how core it is, I would very much appreciate it! Unless its urgent to get merged, I would also look once I get time as this implements some of the things @masnax and I discussed (I think).

**Comment by tomponline** [comment]
@masnax this is going to need a rebase...

**Comment by masnax** [comment]
> @masnax this is going to need a rebase...

rebased.

File: lxd/images.go [hunk 1]
```diff
@@ -2131,7 +2131,6 @@ func pruneExpiredImages(ctx context.Context, d *Daemon, op *operations.Operation
 }
 
 func pruneExpiredImagesInProject(ctx context.Context, d *Daemon, project db.Project, config map[string]string, op *operations.Operation) error {
```
Comment (tomponline) [inline lxd/images.go:2133]: @masnax this looks like it can be sent up as a standalone PR, if possible can you do this so we can get this PR merged in smaller chunks. Thanks

File: lxd/storage/backend_lxd.go [hunk 2]
```diff
@@ -2769,6 +2769,55 @@ func (b *lxdBackend) DeleteImage(fingerprint string, op *operations.Operation) e
 	return nil
 }
 
+// DeleteImageTx removes an image from the database and underlying storage device if needed.
+// TODO: Replace DeleteImage with this method.
+func (b *lxdBackend) DeleteImageTx(fingerprint string, tx *db.ClusterTx, op *operations.Operation) error {
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:2774]: @masnax as deleting volumes from a storage pool can be slow I would definitely not be confident that running multiple of these operations inside a single transaction would be a good idea, as it would likely hold up the whole database from having writes performed by other queries whilst this is in effect (sqlite is write locked at the db level).

File: shared/api/storage_pool_volume.go [hunk 3]
```diff
@@ -120,6 +122,12 @@ type StorageVolume struct {
 	ContentType string `json:"content_type" yaml:"content_type"`
 }
 
+// URL returns the URL for the storage volume.
+func (storageVolume *StorageVolume) URL(apiVersion string, pool string, project string, target string) *URL {
+	volPath := fmt.Sprintf("%s/volumes/%s/%s", pool, storageVolume.Type, storageVolume.Name)
```
Comment (tomponline) [inline shared/api/storage_pool_volume.go:127]: This commit can be a separate PR.

Also, why are you building volPath in an unescaped way rather than passing the 3 params to `Path()` function below it?

File: lxd/api_internal.go [hunk 4]
```diff
@@ -895,8 +895,8 @@ func internalImportRootDevicePopulate(instancePoolName string, localDevices map[
 	expandedRootName, expandedRootConfig, _ := shared.GetRootDiskDevice(expandedDevices)
```
Comment (tomponline) [inline lxd/api_internal.go:895]: Can this commit ` lxd/db/profiles: Use plain map for ExpandInstanceDevices ` be in a separate PR?

**Review by tomponline — COMMENTED** [review]
Overall this looks like its on the right track though! Thanks :)

**Comment by masnax** [comment]
@tomponline One possible way I can make this PR a bit easier to digest is to remove the entire generation component for a while. 

What I mean is I can split this PR into the following multiple PRs:
* I can take the generated methods and add them as new, non-generated helpers as needed. (Maybe in some sort of `<entity>.mapper_v2.go` files)
* Introduce new PRs using these new helpers, we can 1-by-1 go through all the different entities (Instances, projects, profiles, etc) and apply the single-query-per-transaction rule to their usages where appropriate. Potentially some of the simpler ones could be aggregated once the overall plan becomes more clear.
* With everything transitioned to the new helpers, a PR with the code-generation logic can be merged, which will clean out all of the now-unused pre-refactor generated methods, and delete the placeholder `mapper_v2` files for the new methods.

What do you think?

**Comment by tomponline** [comment]
> @tomponline One possible way I can make this PR a bit easier to digest is to remove the entire generation component for a while.
> 
> What I mean is I can split this PR into the following multiple PRs:
> 
>     * I can take the generated methods and add them as new, non-generated helpers as needed. (Maybe in some sort of `<entity>.mapper_v2.go` files)
> 
>     * Introduce new PRs using these new helpers, we can 1-by-1 go through all the different entities (Instances, projects, profiles, etc) and apply the single-transction rule to their usages where appropriate. Potentially some of the simpler ones could be aggregated once the overall plan becomes more clear.
> 
>     * With everything transitioned to the new helpers, a PR with the code-generation logic can be merged, which will clean out all of the now-unused pre-refactor generated methods, and delete the placeholder `mapper_v2` files for the new methods.
> 
> 
> What do you think?

Yes this sounds like it would work. So basically do a few manually at first to try it out?

**Comment by tomponline** [comment]
@masnax I'll close this one now as we're going down the multiple smaller PR route.
